signac/contrib/_slim_collection.py

```diff
@@ -0,0 +1,295 @@
+# Copyright (c) 2022 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""An optimized _SlimCollection class for indexing signac Projects."""
+
+import json
+import logging
+from numbers import Number
+
+from ..errors import InvalidKeyError
+from .collection import (
+    _INDEX_OPERATORS,
+    _check_logical_operator_argument,
+    _DictPlaceholder,
+    _find_with_index_operator,
+    _float,
+    _TypedSetDefaultDict,
+    _valid_filter,
+)
+from .utility import _nested_dicts_to_dotted_keys, _to_hashable
+
+logger = logging.getLogger(__name__)
+
+_PRIMARY_KEY = "_id"
+
+
+class _SlimCollection(dict):
```

Comment:
@csadorf I appreciate these suggestions! I had a similar series of ideas -- I kept `_SlimCollection` for the initial round review partly because I knew the name should change but didn't want to commit to a name until reviewing its use cases and getting feedback like what you provided. I lean towards `_Index`, which should not conflict with any other internals now that we've removed other forms of indexing from the `next` branch. My only hesitation is that it has a method `build_index`, which is an index organized by key. I wonder if that method (or the class) should be named differently. Otherwise the result of `_Index.build_index` is just a `dict` with a particular structure, while it sounds like it should be a class factory method.

Reply:
For now, I've renamed this class to `_Index`.

Reply:
I would strongly prefer a more disambiguated name. While I agree that `_Index` is not currently conflicting with anything, I don't want to hobble our ability to use related names in the future. Otherwise we could run into the same issues that we now have in signac-flow where vague naming leads to ambiguity between concepts like groups, aggregates, bundles, etc. Having reviewed the code, since the critical operation is the ability to search the elements maybe it's something like a `_SearchableIndex`? I would definitely like to avoid the term `Collection` given a) the overloads associated with collections.abc (and now SyncedCollections) and b) we are no longer trying to specifically make this class behave so much like a MongoDB collection